File: backend/python-utils/openrouter_vision.py
#!/usr/bin/env python3
"""
OpenRouter Vision API - Test free vision models
"""
import requests
import base64
import sys
import os

# Configuration
OPENROUTER_API_KEY = "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"
OPENROUTER_URL = "https://openrouter.ai/api/v1/chat/completions"
SITE_URL = "https://softaware.net.za"  # Required by OpenRouter
APP_NAME = "Vision Analysis Tool"  # Required by OpenRouter

# Free vision models on OpenRouter
MODELS = {
    "gemini": "google/gemini-2.0-flash-exp:free",
    "qwen-vl": "qwen/qwen-2-vl-7b-instruct:free",
    "llama-vision": "meta-llama/llama-3.2-11b-vision-instruct:free"
}

def encode_image(image_path):
    """Encode image to base64"""
    with open(image_path, "rb") as image_file:
        return base64.b64encode(image_file.read()).decode('utf-8')

def analyze_image(image_path, model_key="qwen-vl", prompt="Describe what you see in this image. Be concise and accurate."):
    """
    Analyze an image using OpenRouter vision models
    
    Args:
        image_path: Path to the image file
        model_key: Model to use (gemini, qwen-vl, llama-vision)
        prompt: Custom prompt for analysis
    
    Returns:
        str: Model response or error message
    """
    if not os.path.exists(image_path):
        return f"Error: Image file not found: {image_path}"
    
    if model_key not in MODELS:
        return f"Error: Invalid model. Choose from: {', '.join(MODELS.keys())}"
    
    model = MODELS[model_key]
    
    try:
        # Encode image
        base64_image = encode_image(image_path)
        
        # Prepare request
        headers = {
            "Authorization": f"Bearer {OPENROUTER_API_KEY}",
            "HTTP-Referer": SITE_URL,
            "X-Title": APP_NAME,
            "Content-Type": "application/json"
        }
        
        payload = {
            "model": model,
            "messages": [
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "text",
                            "text": prompt
                        },
                        {
                            "type": "image_url",
                            "image_url": {
                                "url": f"data:image/jpeg;base64,{base64_image}"
                            }
                        }
                    ]
                }
            ]
        }
        
        # Send request
        response = requests.post(OPENROUTER_URL, headers=headers, json=payload, timeout=30)
        response.raise_for_status()
        
        result = response.json()
        
        if "choices" in result and len(result["choices"]) > 0:
            return result["choices"][0]["message"]["content"]
        else:
            return f"Unexpected response format: {result}"
            
    except requests.exceptions.Timeout:
        return "Error: Request timed out"
    except requests.exceptions.RequestException as e:
        return f"Request error: {e}"
    except Exception as e:
        return f"Error: {e}"

def main():
    """CLI interface"""
    if len(sys.argv) < 2:
        print("Usage: python3 openrouter_vision.py <image_path> [model] [prompt]")
        print(f"\nAvailable models:")
        for key, value in MODELS.items():
            print(f"  {key}: {value}")
        print(f"\nDefault: google-flash")
        sys.exit(1)
    
    image_path = sys.argv[1]
    model_key = sys.argv[2] if len(sys.argv) > 2 else "qwen-vl"
    prompt = sys.argv[3] if len(sys.argv) > 3 else "Describe what you see in this image. Be concise and accurate."
    
    print(f"Analyzing: {image_path}")
    print(f"Model: {MODELS.get(model_key, 'unknown')}")
    print("-" * 60)
    
    result = analyze_image(image_path, model_key, prompt)
    print(result)
    print("-" * 60)

if __name__ == "__main__":
    main()

File: backend/src/services/openRouterVision.js
import { getSecret } from './credentialVault.js';
// Resolved lazily on first call — reads from encrypted credential vault.
// NEVER hardcode the key.
let _cachedKey = null;
async function getApiKey() {
    if (_cachedKey)
        return _cachedKey;
    _cachedKey = await getSecret('OPENROUTER');
    // Clear cache after 5 min so vault rotation takes effect
    setTimeout(() => { _cachedKey = null; }, 5 * 60 * 1000);
    return _cachedKey;
}
const OPENROUTER_URL = 'https://openrouter.ai/api/v1/chat/completions';
const OPENROUTER_MODELS = [
    'qwen/qwen-2.5-vl-7b-instruct',
    'meta-llama/llama-3.2-11b-vision-instruct',
    'openai/gpt-4o-mini',
];
const SITE_URL = 'https://softaware.net.za';
const APP_NAME = 'Softaware Vision Analysis';
async function callOpenRouter(params) {
    // Build OpenRouter message format
    const openRouterMessages = params.messages.map((msg) => {
        if (!msg.images || msg.images.length === 0) {
            return {
                role: msg.role,
                content: msg.content,
            };
        }
        // Message with images
        const contentParts = [
            {
                type: 'text',
                text: msg.content,
            },
        ];
        // Add images
        for (const img of msg.images) {
            contentParts.push({
                type: 'image_url',
                image_url: {
                    url: `data:${img.mimeType};base64,${img.dataBase64}`,
                },
            });
        }
        return {
            role: msg.role,
            content: contentParts,
        };
    });
    const apiKey = await getApiKey();
    if (!apiKey)
        throw new Error('OpenRouter API key not configured — check credential vault');
    const response = await fetch(OPENROUTER_URL, {
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${apiKey}`,
            'HTTP-Referer': SITE_URL,
            'X-Title': APP_NAME,
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            model: params.model,
            messages: openRouterMessages,
            temperature: params.temperature ?? 0.7,
            max_tokens: params.maxTokens ?? 1024,
        }),
        signal: AbortSignal.timeout(30000), // 30 second timeout
    });
    if (!response.ok) {
        const errorText = await response.text().catch(() => '');
        const error = Object.assign(new Error(`OpenRouter API error: ${response.status} ${errorText}`.trim()), { status: response.status });
        throw error;
    }
    const data = await response.json();
    if (!data.choices || data.choices.length === 0) {
        throw new Error('No response from OpenRouter');
    }
    return {
        content: data.choices[0].message.content,
        model: data.model || params.model,
    };
}
export async function analyzeWithOpenRouter(params) {
    try {
        console.log('[OpenRouter] Starting vision analysis, messages:', params.messages.length);
        let lastError;
        for (const model of OPENROUTER_MODELS) {
            try {
                return await callOpenRouter({
                    model,
                    messages: params.messages,
                    temperature: params.temperature,
                    maxTokens: params.maxTokens,
                });
            }
            catch (error) {
                lastError = error;
                const status = error?.status;
                if (status && status !== 429 && status < 500) {
                    break;
                }
            }
        }
        throw lastError || new Error('OpenRouter request failed');
    }
    catch (error) {
        console.error('OpenRouter vision error:', error);
        throw error;
    }
}
